### PR TITLE
Allow even more date formats

### PR DESCRIPTION
**Pull request by aug24**
## What does this change?
We have found yet another form of date which is sent in exif data: `yyyyMMdd`.

This is added to the list of tolerated date formats (in the right place).  At the same time, a few other likely candidates were added, plus tests for them.

## How can success be measured?
Fewer errors on load.

## Tested?
- [ ] locally by committer
- [ ] locally by Guardian reviewer
- [ ] on the Guardian's TEST environment

### CODE REVIEW

**Review by paperboyo — APPROVED** [review]
This will help not only recognise these (illegal by spec!) dates, but will also stop failing the ingestion of images carrying them too. This latter point should be guarded against separately, obviously. And then, the failed images in s3 (which do not exist in ES) reingested. Internal [ticket here](https://trello.com/c/qwohLB8Y/2158-fail-to-index-stores-s3-objects-missing-from-es-beacuse-of-a-daft-date).

File: common-lib/src/main/scala/com/gu/mediaservice/lib/metadata/ImageMetadataConverter.scala [hunk 1]
```diff
@@ -126,6 +126,9 @@ object ImageMetadataConverter extends GridLogging {
     DateTimeFormat.forPattern("E MMM dd HH:mm:ss.SSS 'BST' yyyy").withZone(DateTimeZone.forOffsetHours(1)),
     DateTimeFormat.forPattern("E MMM dd HH:mm:ss 'BST' yyyy").withZone(DateTimeZone.forOffsetHours(1)),
 
+    DateTimeFormat.forPattern("yyyyMMdd"),
+    DateTimeFormat.forPattern("yyyyMM"),
+    DateTimeFormat.forPattern("yyyyddMM"),
```
Comment (AWare) [inline common-lib/src/main/scala/com/gu/mediaservice/lib/metadata/ImageMetadataConverter.scala:131]: I'm not sure whether we want to try both of these unless we've seen them, as they'll overlap for some dates- and we'll need to thandle that.

Comment (paperboyo) [reply common-lib/src/main/scala/com/gu/mediaservice/lib/metadata/ImageMetadataConverter.scala:131]: I thought they are tried from the top. So we are (correctly) Europe-centric (and ambigous ones are `yyyMMdd`-first). In case that’s not true, you are right.
Currently, there is no way to tell what is in the wild in those (any) fields. But after glorious [this](https://github.com/guardian/grid/issues/3100) is in (and we make it accept wildcards), we will be able to (finally) tell.
Date parsing logic could be much smarter than currently (comparing, looking for oldest and most detailed), but this is a good step regardless.

Comment (aug24) [reply common-lib/src/main/scala/com/gu/mediaservice/lib/metadata/ImageMetadataConverter.scala:131]: They are indeed tried from the top, fail-down.  So that's OK, we handle that by simple ordering, and there are tests to prove it too.

Comment (aug24) [reply common-lib/src/main/scala/com/gu/mediaservice/lib/metadata/ImageMetadataConverter.scala:131]: NB the yyyyMM has to come before yyyy for precisely this reason.  Any corner cases which are suggested will be added to tests...??

Comment (AWare) [reply common-lib/src/main/scala/com/gu/mediaservice/lib/metadata/ImageMetadataConverter.scala:131]: I'm not super happy about American dates of this format silently changing for the first 12 days of the month and then are parsed correctly. I'd much rather have them silently succeed for 12 days and fail the rest of the month until we're sure they exist; at which point we can figure out how to distinguish them.

Comment (paperboyo) [reply common-lib/src/main/scala/com/gu/mediaservice/lib/metadata/ImageMetadataConverter.scala:131]: Just to note: these dates were observed only as `yyyyMMdd` in both `photoshop:DateCreated` and `iptc:Date Created` (which is [mixed by metadata-extractor](https://github.com/guardian/grid/blob/ddb9effd49fa39bf63bd7e3bcd0fad3af78a19f1/image-loader/app/lib/imaging/FileMetadataReader.scala#L97-L105) into `Date Time Created Composite`. They are **illegal** values for both of those fields. We are doing them a favour by not dropping them on the floor. Instead of trying to figure out if this is silly American or correct European format, which may not be ultimately possible, we could improve other areas of date parsing (starting with not failing to ingest or storing in s3, but not in ES). Or, might I say, starting with virtually anything else ;-)

Comment (sihil) [reply common-lib/src/main/scala/com/gu/mediaservice/lib/metadata/ImageMetadataConverter.scala:131]: Can we set up an auto feedback loop that emails suppliers when they send us images with illegal dates?

Comment (paperboyo) [reply common-lib/src/main/scala/com/gu/mediaservice/lib/metadata/ImageMetadataConverter.scala:131]: > Can we set up an auto feedback loop that emails suppliers when they send us images with illegal dates?

Without much exaggeration, I’m afraid, for most big agencies, it would be an email for every image, if we would to do it. 😜 And the examples of these came via `email` route, so we would need to build tech that preserves the sender in metadata.

Look, this isn’t ideal and I could definitely be less facetious about the American date format. We can remove it from the above subprime position, but then we would need to somehow deal with it when it comes in anyway, right? Sniffing possible formatting sounds near impossible.

And all of it is just more code, for the examples that we haven’t encountered, while the ones that we have are trickling into the Grid as I type this too long comment **and they literally end up in s3 only, without any record in ES** generating upload errors. I would rather spend any effort into looking at why we fail to ingest an image, because some metadata fields has some illegal value. Illegal or poor metadata is, sadly, part of life, as our cleaner well illustrate.

Comment (AWare) [inline common-lib/src/main/scala/com/gu/mediaservice/lib/metadata/ImageMetadataConverter.scala:131]:
```suggestion
```

Comment (paperboyo) [reply common-lib/src/main/scala/com/gu/mediaservice/lib/metadata/ImageMetadataConverter.scala:131]: ^ as I commented above, I’m not against. But what does this change mean for a “date” that is `20202201`? (pretty sure it errors and errors badly: s3-only, no ES record!)

Comment (AWare) [reply common-lib/src/main/scala/com/gu/mediaservice/lib/metadata/ImageMetadataConverter.scala:131]: That issue would be better fixed by dropping the field instead of the image.

Comment (paperboyo) [reply common-lib/src/main/scala/com/gu/mediaservice/lib/metadata/ImageMetadataConverter.scala:131]: > That issue would be better fixed by dropping the field instead of the image.

Goes without saying! And without me saying that it goes without saying! :-) But! Your change means we will drop them, so the question is: would we rather pretend there is no date or have a date wrong by half a year?

**Review by twrichards — APPROVED** [review]
LGTM - based on the code, not tested it myself but based on the lovely tests 👏 ... I reckon it's fine
